boot/aether/src/boot/aether.clj

```diff
@@ -59,13 +64,82 @@
+        boot-home (or (and boot-home (io/file boot-home))
+                      (io/file (System/getProperty "user.home") ".boot"))]
+    (.getAbsolutePath (doto boot-home .mkdirs))))
+
+(defn ^{:boot/from :technomancy/leiningen} credentials-fn
+  "Decrypt map from credentials.edn.gpg in Boot home if present."
+  ([] (let [cred-file (io/file (boot-home) "credentials.edn.gpg")]
+        (if (.exists cred-file)
+          (credentials-fn cred-file))))
+  ([file]
+   (let [{:keys [out err exit]} (gpg/gpg "--quiet" "--batch"
+                                         "--decrypt" "--" (str file))]
+     (if (pos? exit)
+       (binding [*out* *err*]
+         (util/warn (format (str "Could not decrypt credentials from %s\n"
+                                 "%s\n"
+                                 "See `boot gpg --help` for how to install gpg.")
+                            (str file) err)))
+       (read-string out)))))
+
+(def credentials (memoize credentials-fn))
+
+(defn- ^{:boot/from :technomancy/leiningen} match-credentials [settings auth-map]
+  (get auth-map (:url settings)
+       (first (for [[re? cred] auth-map
+                    :when (and (instance? Pattern re?)
+                               (re-find re? (:url settings)))]
+                cred))))
+
+(defn- ^{:boot/from :technomancy/leiningen} resolve-credential
+  "Resolve key-value pair from result into a credential, updating result."
+  [source-settings result [k v]]
+  (letfn [(resolve [v]
+            (cond (= :env v)
+                  (System/getenv (str "BOOT_" (string/upper-case (name k))))
+
+                  (and (keyword? v) (= "env" (namespace v)))
+                  (System/getenv (string/upper-case (name v)))
+
+                  (= :gpg v)
+                  (get (match-credentials source-settings (credentials)) k)
+
+                  (coll? v) ;; collection of places to look
+                  (->> (map resolve v)
+                       (remove nil?)
+                       first)
+
+                  :else v))]
+    (if (#{:username :password :passphrase :private-key-file} k)
+      (assoc result k (resolve v))
+      (assoc result k v))))
+
+(defn ^{:boot/from :technomancy/leiningen} resolve-credentials
+  "Applies credentials from the environment or ~/.boot/credentials.edn.gpg
+  as they are specified and available."
+  [settings]
+  (let [gpg-creds (if (= :gpg (:creds settings))
+                    (match-credentials settings (credentials)))
+        resolved (reduce (partial resolve-credential settings)
+                         (empty settings)
+                         settings)]
+    (if gpg-creds
+      (dissoc (merge gpg-creds resolved) :creds)
+      resolved)))
+
 (defn resolve-dependencies*
   [env]
   (try
     (aether/resolve-dependencies
       :coordinates       (:dependencies env)
       :repositories      (->> (or (:repositories env) @default-repositories)
-                           (map (juxt first (fn [[x y]] (if (map? y) y {:url y}))))
```

Comment:
I dropped suport for defining repositories as string instead of map with `:url` property. Could be still added back.